xds/src/main/java/com/linecorp/centraldogma/xds/internal/ControlPlanePlugin.java

```diff
@@ -67,219 +68,236 @@ public final class ControlPlanePlugin extends AllReplicasPlugin {
-                break;
-            case SECRET:
-                secrets = (SnapshotResources<Secret>) resources;
-                break;
-            default:
-                // Should never reach here.
-                throw new Error();
-        }
+    private void handleDiff(Repository repository, Revision newRevision,
+                            CompletableFuture<Map<String, Change<?>>> diffFuture, Revision lastKnownRevision) {
+        diffFuture.handleAsync((BiFunction<Map<String, Change<?>>, Throwable, Void>) (changes, cause) -> {
+            if (stop) {
+                return null;
+            }
+            final String repoName = repository.name();
+            if (cause != null) {
+                logger.warn("Unexpected exception while diffing {} from {} to {}. Building from the first.",
+                            repoName, lastKnownRevision, newRevision, cause);
+                centralDogmaXdsResources.removeProject(repoName);
+                // Do not call cache.setSnapshot(). Let watchXdsProject() create a new snapshot.
+                watchXdsProject(repository, Revision.INIT);
+                return null;
+            }
 
-        cache.setSnapshot(DEFAULT_GROUP,
-                          new CentralDogmaSnapshot(clusters, endpoints, listeners, routes, secrets));
+            for (Change<?> change : changes.values()) {
+                final String path = change.path();
+                switch (change.type()) {
+                    case UPSERT_JSON:
+                        try {
+                            setXdsResources(path, change.contentAsText(), repoName);
+                        } catch (Throwable t) {
+                            logger.warn("Unexpected exception while building an xDS resource from {}.",
+                                        repoName + path, t);
+                        }
+                        break;
+                    case REMOVE:
+                        if (path.startsWith(CLUSTERS_DIRECTORY)) {
+                            centralDogmaXdsResources.removeCluster(repoName, path);
+                        } else if (path.startsWith(ENDPOINTS_DIRECTORY)) {
+                            centralDogmaXdsResources.removeEndpoint(repoName, path);
+                        } else if (path.startsWith(LISTENERS_DIRECTORY)) {
+                            centralDogmaXdsResources.removeListener(repoName, path);
+                        } else if (path.startsWith(ROUTES_DIRECTORY)) {
+                            centralDogmaXdsResources.removeRoute(repoName, path);
+                        }
+                        break;
+                    default:
+                        // Ignore other types of changes.
+                        // No APPLY_JSON_PATCH because the diff option.
+                        // No RENAME because the resource name in the content always have to be
+                        // changed if the file is renamed.
```

Comment:
Question) Is it possible to at least log with warning level here since we can't guarantee this in compile time?

Reply:
That's a good idea. 👍